include/tvm/relay/pass.h

```diff
@@ -2,18 +2,202 @@
+ */
+class PassContextNode : public RelayNode {
+ public:
+  /*!
+   * \brief The error reporter used to notify users why an optimization fails.
+   */
+  ErrorReporter err_reporter_;
+
+  PassContextNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+  }
+
+  TVM_DLL static PassContext make();
+
+  static constexpr const char* _type_key = "relay.PassContext";
+  TVM_DECLARE_NODE_TYPE_INFO(PassContextNode, RelayNode);
+};
+
+class PassContext : public NodeRef {
+ public:
+  PassContext() = default;
+  explicit PassContext(NodePtr<tvm::Node> p) : NodeRef(p) {}
+
+  const PassContextNode* operator->() const {
+    return static_cast<PassContextNode*>(this->node_.get());
+  }
+
+  using ContainerType = PassContextNode;
+};
+
+// We use currying here. It runs on a Relay node type NodeT and yields a new
+// node with the same type. The Relay module is captured for optimizations as
+// most of the current Relay optimizations are module to module. Currying
+// sketches the optimization, i.e. how we want to mutate an AST, and it is
+// passed as packed functions that will be invoked when called by `run`.
+//
+// For example, PassFunc<Function> indicates we perform a Function to Function
+// transformation on the given Module.
+template <typename NodeT,
+          typename = std::enable_if<(std::is_same<NodeT, Module>::value ||
+                                     std::is_same<NodeT, Function>::value)>>
+using PassFunc =
+    runtime::TypedPackedFunc<runtime::TypedPackedFunc<NodeT(NodeT)>(
+        const Module& mod)>;
+
+class Pass;
+
+/*!
+ * \brief PassNode is the base type of differnt types of optimization passes.
+ * It is implemented by different pass subclasses at different granularity of
+ * Relay nodes.
+ */
+class PassNode : public RelayNode {
+ public:
+  /*! \brief The name of an optimization/analysis pass. */
+  std::string name;
+  /*! \brief The minimal optimization level that this pass will be enabled. */
+  int opt_level;
+
+  /*!
+   * \brief Set the context information for a pass.
+   *
+   * \param pass_ctx The context information for a certain pass.
+   */
+  virtual void SetContext(const PassContext& pass_ctx) = 0;
+
+  /*!
+   * \brief Get the required passes for this pass as a vector of std::string.
+   */
+  virtual std::vector<std::string> Required() const = 0;
+
+  /*!
+   * \brief Execute the optimization pass using a functor. This functor invokes
+   *        the `run` method to perform a real optimization on a certain type
+   *        of node.
+   *
+   * \param mod The module that an optimization pass runs on.
+   *
+   * \return The updated module.
+   */
+  virtual Module operator()(const Module& mod) const = 0;
+
+  void VisitAttrs(tvm::AttrVisitor* v) override {
+    v->Visit("name", &name);
+    v->Visit("opt_level", &opt_level);
+  }
+
+  static constexpr const char* _type_key = "relay.Pass";
+  TVM_DECLARE_BASE_NODE_INFO(PassNode, RelayNode);
+};
+
+class Pass : public NodeRef {
```

Comment:
TVM_DEFINE_NODE_REF(Pass, PassNode)

Reply:
We probably need this because we need "SetContext" otherwise it would return `const PassNode*` from `operator->()`, right?

Reply:
OK, fine, we can keep this one as it is